File: 0143-reorder-list/0143-reorder-list.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  void reorderList(ListNode? head) {
    ListNode? slow = head, fast = head;
    while(slow != null && fast != null && fast.next != null){
        slow = slow.next;
        fast = fast.next!.next;
    }

    ListNode? start = slow!.next;
    slow.next = null;

    ListNode? prev = null, curr = start;
    while(curr != null){
        ListNode? next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }

    ListNode? h = head;
    while(h != slow && h != null && prev != null){
        ListNode? h_next = h.next;
        ListNode? prev_next = prev.next;
        h.next = prev;
        prev.next = h_next;
        prev = prev_next;
        h = h_next;
    }
  }
}

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        #lets go the middle
        slow = head
        fast = head

        while slow.next and fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        #reverse from middle
        prev = None
        curr = slow.next
        slow.next = None

        while curr:
            nxt = curr.next
            curr.next = prev

            prev = curr
            curr = nxt

        #merging both parts
        first = head
        last = prev

        while last:
            first_nxt = first.next
            last_nxt = last.next

            first.next = last
            last.next = first_nxt

            first = first_nxt
            last = last_nxt


        